File: Source/JavaScript/model/packages/PackageDependency.ts
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

/**
 * One package a {@link ScenePackage} needs present in the same {@link UiProfile} to work.
 * `@cratis/components` is written against PrimeReact and Tailwind, so it declares both - a profile
 * listing it without them is stating something that cannot render, and that has to be visible rather
 * than discovered at runtime.
 */
export interface PackageDependency {
    /**
     * The name of the package depended on, matching that package's `name`.
     */
    name: string;

    /**
     * An optional semver range the dependency must satisfy. Left undefined when any version will do -
     * which is the common case, since a profile only ever activates one version of a package.
     */
    versionRange?: string;
}

export const PackageDependencyPropertyNames: (keyof PackageDependency)[] = ['name', 'versionRange'];

File: Source/DotNET/Model/Packages/PackageDependency.cs
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Cratis.Scene.Model.Packages;

/// <summary>
/// One package a <see cref="ScenePackage"/> needs present in the same <see cref="Profiles.UiProfile"/> to
/// work. <c>@cratis/components</c> is written against PrimeReact and Tailwind, so it declares both - a
/// profile listing it without them is stating something that cannot render, and that has to be visible
/// rather than discovered at runtime.
/// </summary>
/// <param name="Name">The name of the package depended on, matching that package's <see cref="ScenePackage.Name"/>.</param>
/// <param name="VersionRange">
/// An optional semver range the dependency must satisfy. Left <see langword="null"/> when any version
/// will do - which is the common case, since a profile only ever activates one version of a package.
/// </param>
public record PackageDependency(string Name, string? VersionRange = null);
